File: vane-core/src/main/java/org/oddlama/vane/core/module/Context.java
package org.oddlama.vane.core.module;

import java.io.IOException;
import java.util.function.Consumer;
import org.bukkit.scheduler.BukkitTask;
import org.json.JSONObject;
import org.oddlama.vane.core.functional.Consumer1;
import org.oddlama.vane.core.resourcepack.ResourcePackGenerator;

/**
 * A ModuleContext is an association to a specific Module and also a grouping of config and language
 * variables with a common namespace.
 */
public interface Context<T extends Module<T>> {
    public static String appendYamlPath(String ns1, String ns2, String separator) {
        if (ns1.isEmpty()) {
            return ns2;
        }
        return ns1 + separator + ns2;
    }

    /** create a sub-context namespace */
    public default ModuleContext<T> namespace(String name) {
        return new ModuleContext<T>(this, name, null, ".");
    }

    /** create a sub-context namespace */
    public default ModuleContext<T> namespace(String name, String description) {
        return new ModuleContext<T>(this, name, description, ".");
    }

    /** create a sub-context namespace */
    public default ModuleContext<T> namespace(String name, String description, String separator) {
        return new ModuleContext<T>(this, name, description, separator);
    }

    /** create a sub-context group */
    public default ModuleGroup<T> group(String group, String description) {
        return new ModuleGroup<T>(this, group, description);
    }

    public default ModuleGroup<T> group(String group, String description, boolean defaultEnabled) {
        final var g = new ModuleGroup<T>(this, group, description);
        g.configEnabledDef = defaultEnabled;
        return g;
    }

    /** create a sub-context group */
    public default ModuleGroup<T> groupDefaultDisabled(String group, String description) {
        final var g = group(group, description);
        g.configEnabledDef = false;
        return g;
    }

    /**
     * Compile the given component (processes lang and config definitions) and registers it for
     * onModuleEnable, onModuleDisable and onConfigChange events.
     */
    public void compile(ModuleComponent<T> component);

    public void addChild(Context<T> subcontext);

    public Context<T> getContext();

    public T getModule();

    public String yamlPath();

    public String variableYamlPath(String variable);

    public boolean enabled();

    public void enable();

    public void disable();

    public void configChange();

    public void generateResourcePack(final ResourcePackGenerator pack) throws IOException;

    public void forEachModuleComponent(final Consumer1<ModuleComponent<?>> f);

    public default void onModuleEnable() {}

    public default void onModuleDisable() {}

    public default void onConfigChange() {}

    public default void onGenerateResourcePack(final ResourcePackGenerator pack) throws IOException {}

    public default BukkitTask scheduleTaskTimer(Runnable task, long delayTicks, long periodTicks) {
        return getModule().getServer().getScheduler().runTaskTimer(getModule(), task, delayTicks, periodTicks);
    }

    public default BukkitTask scheduleTask(Runnable task, long delayTicks) {
        return getModule().getServer().getScheduler().runTaskLater(getModule(), task, delayTicks);
    }

    public default BukkitTask scheduleNextTick(Runnable task) {
        return getModule().getServer().getScheduler().runTask(getModule(), task);
    }

    public default void addStorageMigrationTo(long to, String name, Consumer<JSONObject> migrator) {
        getModule().persistentStorageManager.addMigrationTo(to, name, migrator);
    }

    public default String storagePathOf(String field) {
        if (!field.startsWith("storage")) {
            throw new RuntimeException("Configuration fields must be prefixed storage. This is a bug.");
        }
        return variableYamlPath(field.substring("storage".length()));
    }

    public default void markPersistentStorageDirty() {
        getModule().markPersistentStorageDirty();
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/module/Context.kt
package org.oddlama.vane.core.module

import org.bukkit.scheduler.BukkitTask
import org.json.JSONObject
import org.oddlama.vane.core.functional.Consumer1
import org.oddlama.vane.core.resourcepack.ResourcePackGenerator
import java.io.IOException
import java.util.function.Consumer

/**
 * A ModuleContext is an association to a specific Module and also a grouping of config and language
 * variables with a common namespace.
 */
interface Context<T : Module<T?>?> {
    /**
     * Creates a sub-context namespace.
     */
    @Suppress("UNCHECKED_CAST")
    fun namespace(name: String?): ModuleContext<T?> {
        val ctx: Context<T?> = this as Context<T?>
        return ModuleContext(ctx, name, null, ".")
    }

    /**
     * Creates a described sub-context namespace.
     */
    @Suppress("UNCHECKED_CAST")
    fun namespace(name: String?, description: String?): ModuleContext<T?> {
        val ctx: Context<T?> = this as Context<T?>
        return ModuleContext(ctx, name, description, ".")
    }

    /**
     * Creates a described sub-context namespace with a custom separator.
     */
    @Suppress("UNCHECKED_CAST")
    fun namespace(name: String?, description: String?, separator: String?): ModuleContext<T?> {
        val ctx: Context<T?> = this as Context<T?>
        return ModuleContext(ctx, name, description, separator)
    }

    /**
     * Creates a sub-context group with a configurable enabled flag.
     */
    @Suppress("UNCHECKED_CAST")
    fun group(group: String?, description: String?): ModuleGroup<T?> {
        val ctx: Context<T?> = this as Context<T?>
        return ModuleGroup.create(ctx, group, description)
    }

    /**
     * Creates a sub-context group and sets its default enabled state.
     */
    @Suppress("UNCHECKED_CAST")
    fun group(group: String?, description: String?, defaultEnabled: Boolean): ModuleGroup<T?> {
        val ctx: Context<T?> = this as Context<T?>
        val g = ModuleGroup.create(ctx, group, description)
        g.configEnabledDef = defaultEnabled
        return g
    }

    /**
     * Creates a sub-context group that is disabled by default.
     */
    fun groupDefaultDisabled(group: String?, description: String?): ModuleGroup<T?> {
        val g = group(group, description)
        g.configEnabledDef = false
        return g
    }

    /**
     * Compile the given component (processes lang and config definitions) and registers it for
     * onModuleEnable, onModuleDisable and onConfigChange events.
     */
    fun compile(component: ModuleComponent<T?>?)

    /**
     * Adds a child context.
     */
    fun addChild(subcontext: Context<T?>?)

    /**
     * Returns the parent context.
     */
    val context: Context<T?>?

    /**
     * Returns the owning module.
     */
    val module: T?

    /**
     * Returns this context YAML path.
     */
    fun yamlPath(): String?

    /**
     * Returns the YAML path for a variable under this context.
     */
    fun variableYamlPath(variable: String?): String?

    /**
     * Returns whether this context is enabled.
     */
    fun enabled(): Boolean

    /**
     * Enables this context and descendants.
     */
    fun enable()

    /**
     * Disables this context and descendants.
     */
    fun disable()

    /**
     * Dispatches a configuration reload event.
     */
    fun configChange()

    /**
     * Generates resource pack contributions for this context tree.
     */
    @Throws(IOException::class)
    fun generateResourcePack(pack: ResourcePackGenerator?)

    /**
     * Visits every registered module component.
     */
    fun forEachModuleComponent(f: Consumer1<ModuleComponent<*>?>?)

    /**
     * Hook called before child/component enable dispatch.
     */
    fun onModuleEnable() {}

    /**
     * Hook called after child/component disable dispatch.
     */
    fun onModuleDisable() {}

    /**
     * Hook called before child/component config-change dispatch.
     */
    fun onConfigChange() {}

    /**
     * Hook called before child/component resource-pack generation dispatch.
     */
    @Throws(IOException::class)
    fun onGenerateResourcePack(pack: ResourcePackGenerator?) {
    }

    /**
     * Schedules a repeating task in the owning module scheduler.
     */
    fun scheduleTaskTimer(task: Runnable, delayTicks: Long, periodTicks: Long): BukkitTask =
        this.module!!.server.scheduler.runTaskTimer(this.module!!, task, delayTicks, periodTicks)

    /**
     * Schedules a delayed task in the owning module scheduler.
     */
    fun scheduleTask(task: Runnable, delayTicks: Long): BukkitTask =
        this.module!!.server.scheduler.runTaskLater(this.module!!, task, delayTicks)

    /**
     * Schedules a task for the next server tick.
     */
    fun scheduleNextTick(task: Runnable): BukkitTask =
        this.module!!.server.scheduler.runTask(this.module!!, task)

    /**
     * Registers a persistent-storage migration step.
     */
    fun addStorageMigrationTo(to: Long, name: String?, migrator: Consumer<JSONObject?>) {
        this.module!!.persistentStorageManager.addMigrationTo(to, name, migrator)
    }

    /**
     * Resolves the persistent-storage path for a storage field.
     */
    fun storagePathOf(field: String): String? {
        require(field.startsWith("storage")) { "Configuration fields must be prefixed storage. This is a bug." }
        return variableYamlPath(field.substring("storage".length))
    }

    /**
     * Marks persistent storage as dirty.
     */
    fun markPersistentStorageDirty() {
        this.module!!.markPersistentStorageDirty()
    }

    /**
     * Utility helpers for context path handling.
     */
    companion object {
        /**
         * Appends a path segment to a namespace using the provided separator.
         */
        @JvmStatic
        fun appendYamlPath(ns1: String, ns2: String?, separator: String?): String? =
            if (ns1.isEmpty()) ns2 else ns1 + separator + ns2
    }
}
